Refactor docs to use ExDocs search_data model for global HexDocs indexing of Gleam packages

diemogebhardt opened the pull request:
This PR is regarding issue https://github.com/gleam-lang/gleam/issues/3811.

## Data Model

The search data model json changed from `SearchIndex`:

```json
[
  {
    "doc": "the name of the module or page",
    "title": "the title of the item",
    "content": "the content of the item",
    "url": "the relative link of the item, like 'module.html#title'"
  }
]
```

to `SearchData`, `SearchItem`, `SearchItemType` and `SearchProgrammingLanguage`:

```json
{
  "items": [
    {
      "type": "module" | "type" | "function" | "constant" | "page",
      "parentTitle": "the name of the module or page",
      "title": "the title of the item",
      "doc": "the content of the item",
      "ref": "the relative link of the item, like 'module.html#title'"
    }
  ],
  "proglang": "gleam"
}
```

Notes:
- `parentTitle` doesn't seem to have an equivalent in ExDocs from what I have seen so far. It is being populated by:
  - either `module.name.to_string()` in the context of types `"module" | "type" | "function" | "constant"`
  - or `config.name.to_string()` in the context of type `"page"` (`README.md`)
- `doc` matches ExDocs, except for headline partitioning
- `ref` matches ExDocs, except for arity `module.html#title/arity`

Neither of those is an issue for global HexDocs indexing of Gleam packages.

## Output Files

~~As I'm still exploring how everything works in the codebase, I didn't want to change too much at once. So I added `search_data.json` in addition to `search-data.js` to keep the existing mechanism working as is for now.~~

### `search-data.js`

~~Wrapped `search_data_json` in `window.Gleam.initSearch({});` for use in `documentation_layout.html` to execute `initSearch({})` in `docs-js/index.js` which in turn uses `lunr.js`.~~

Has been removed in favor of `search_data.json`.

### `search_data.json`

Bare `search_data_json` for use in global HexDocs.

diemogebhardt commented:
Yes, test are a good idea! :)

Lets clarify the other topics first though: whether we keep two separate data models. One for Gleam internal use and another for HexDocs indexing?

lpil commented:
I think just one is fine, it's only used to render that JSON object.

diemogebhardt commented:
Ok, I don't know what you want me to do. This isn't a good use for either your, nor my time.

Not sure how familiar you are with this part of the code base and when you looked into it last. From my own experience, I know that running a project while overseeing a larger code base with lots of different contributors is quite something. So I understand that revisiting parts of the code base that others have contributed to and you haven't reviewed recently takes getting used to. That's why I added context information, including screenshots in the initial post and comments regarding key changes to make it easier to follow. Do you see them or are they just visible to me?

Is there anything else I can do to communicate better?

diemogebhardt commented:
## Open Questions

- `Extra` vs `Page`? `Extra` being the ExDocs/HexDocs way.
- `parentTitle` or does a better name come to mind?
- Is consistent naming throughout the code base and languages a goal? Options:
  - 2 Models:
    - one internal with extended names for rust and js
    - one external with abbreviated names for global HexDocs indexing
  - 1 Model variant A:
    - rust model named with abbreviated names
    - js models with abbreviated names
  - 1 Model variant B:
    - rust model named with extended names
    - js models with abbreviated names
  - ...?


## Current implementation: 1 Model variant A

###  `docs.rs`

```rs
#[derive(Serialize, PartialEq, Eq, PartialOrd, Ord, Clone)]
struct SearchData {
    items: Vec<SearchItem>,
    proglang: SearchProgLang,
}

#[derive(Serialize, PartialEq, Eq, PartialOrd, Ord, Clone)]
struct SearchItem {
    #[serde(rename = "type")]
    type_: SearchItemType,
    #[serde(rename = "parentTitle")]
    parent_title: String,
    title: String,
    doc: String,
    #[serde(rename = "ref")]
    ref_: String,
}

#[derive(Serialize, PartialEq, Eq, PartialOrd, Ord, Clone)]
#[serde(rename_all = "lowercase")]
enum SearchItemType {
    Constant,
    Function,
    Module,
    Page,
    Type,
}

#[derive(Serialize, PartialEq, Eq, PartialOrd, Ord, Clone)]
#[serde(rename_all = "lowercase")]
enum SearchProgLang {
    // Elixir,
    // Erlang,
    Gleam,
}
```

outputs:

### `search-data.js`

for use in [compiler-core/templates/docs-js/index.js](https://github.com/gleam-lang/gleam/pull/3904/files#diff-31b485edc40e2995bbe66fb04f2642abd41d2886d63339760986d8972a2cc07d):

```js
window.Gleam.initSearch({
  "items": [
    {
      "type": "module" | "type" | "function" | "constant" | "page",
      "parentTitle": "the name of the module or page",
      "title": "the title of the item",
      "doc": "the content of the item",
      "ref": "the relative link of the item, like 'module.html#title'"
    }
  ],
  "proglang": "gleam"
});
```

to generate search results in HexDocs of Gleam projects like:

![image](https://github.com/user-attachments/assets/62777238-c70a-4302-b145-4b2328113106)

and

### `search-data.json`

for use in global `HexDocs` indexing:

```json
{
  "items": [
    {
      "type": "module" | "type" | "function" | "constant" | "page",
      "parentTitle": "the name of the module or page",
      "title": "the title of the item",
      "doc": "the content of the item",
      "ref": "the relative link of the item, like 'module.html#title'"
    }
  ],
  "proglang": "gleam"
}
```

lpil commented:
Hello! Sorry for the delay, my inbox is overwhelming!

I'm a bit confused by your notes here. Are you suggesting we propose a change upstream to Hex for them to change the format that all BEAM languages use? Personally I am happy with the format Hex is using so if the comments from my previous PR are applied then we can merge. Thank you

diemogebhardt commented:
I give up, @lpil – our miscommunication is too frustrating.

Above, there are two questions that are still unanswered since the initial post of this PR:
- `Extra` vs `Page`? `Extra` being the `ExDocs`/`HexDocs` way.
  - If `Extra`, only for use in the model or also throughout the rest of the code base?
- `parent_title`/`parentTitle` or does a better name come to mind?
  - used in HexDocs search of Gleam projects
  - **NOT** used in global HexDocs indexing of Gleam projects

Above, there are three options mapped out – including a draft implementation, and their pros and cons. Have you had a look at them and understand the impact of each? These are the three options to choose from:
- Option: 2 Models
- Option: 1 Model variant A (current implementation)
- Option: 1 Model variant B

Feel free to close this unless you plan on addressing the above.

lpil commented:
I would like the specific field renaming that I asked for above please 🙏 I am not looking to discuss any alternative designs presently.

If you would like to do just the tests and leave the renaming to me that would also be fine. Thank you

diemogebhardt commented:
Same here, @lpil – sorry! Glad we figured this out – I am relieved. You should be able to see my comments from weeks back now.

diemogebhardt commented on compiler-core/src/docs.rs:
```diff
@@ -808,9 +821,37 @@ struct ModuleTemplate<'a> {
-    doc: String,
+struct SearchData {
+    items: Vec<SearchItem>,
+    proglang: SearchProgLang,
```
Done

diemogebhardt commented on compiler-core/src/docs.rs:
```diff
@@ -808,9 +821,37 @@ struct ModuleTemplate<'a> {
-    url: String,
+    doc: String,
+    #[serde(rename = "ref")]
+    ref_: String,
```
Done

diemogebhardt commented on compiler-core/src/docs.rs:
```diff
@@ -808,9 +821,37 @@ struct ModuleTemplate<'a> {
     title: String,
-    content: String,
-    url: String,
+    doc: String,
```
Done

diemogebhardt commented on compiler-core/src/docs.rs:
```diff
@@ -808,9 +821,37 @@ struct ModuleTemplate<'a> {
+
+#[derive(Serialize, PartialEq, Eq, PartialOrd, Ord, Clone)]
+#[serde(rename_all = "lowercase")]
+enum SearchProgLang {
```
Done

diemogebhardt commented:
> Thank you!! I've left a few small comments inline, mostly around abbreviations, which we avoid in this codebase.
Renamed the abbreviations. You should be able to see all my replies to your comments from weeks ago now.

> It would be great to have a test for the output format also.
Added tests to ensure that the output matches the ExDocs specification.

diemogebhardt commented:
Rebased and resolved merge conflicts.

lpil commented on compiler-core/src/docs.rs:
```diff
@@ -358,20 +363,27 @@ pub fn generate_html<IO: FileSystemReader>(
-                .expect("search index serialization")
-        )),
+        content: Content::Text(format!("window.Gleam.initSearch({});", search_data_json)),
+    });
```
Remove this please, we don't have to have the same data twice. Everything should use the same JSON file.

diemogebhardt replied:
Done, see [comment below](https://github.com/gleam-lang/gleam/pull/3904#discussion_r1926332587).

lpil reviewed (commented):
Wonderful! I've left one small note inline and then we're good to merge.

Thank you for your patience while I dealt with a family matter.

diemogebhardt commented on compiler-core/templates/documentation_layout.html:
```diff
@@ -302,10 +302,13 @@ <h2>Modules</h2>
-    <!-- Load the search index using JSONP to avoid CORS issues -->
-    <script src="{{ unnest }}/search-data.js?v={{ rendering_timestamp }}"></script>
+    <script>
+      fetch("{{ unnest }}/search_data.json?v={{ rendering_timestamp }}")
```
The response headers include `Access-Control-Allow-Origin: *`, so CORS shouldn't be an issue.

diemogebhardt commented on compiler-core/src/docs.rs:
```diff
@@ -358,20 +363,22 @@ pub fn generate_html<IO: FileSystemReader>(
+    .expect("search index serialization");
+
     files.push(OutputFile {
-        path: Utf8PathBuf::from("search-data.js"),
```
Removed `search-data.js` in favor of `search_data.json`.

diemogebhardt commented:
> Wonderful! I've left one small note inline and then we're good to merge.

Anything else?

> Thank you for your patience while I dealt with a family matter.

No worries, hope everything is ok!

lpil reviewed (approved):
Thank you!!
